server/api/notification.go

```diff
@@ -0,0 +1,261 @@
+		quit:       make(chan (bool)),
+	}
+	go h.work()
+	return h
+}
+
+func (h *notificationHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	notifications := h.Remote.HandleNotification(w, req, h.loadUserSubscription)
+	for _, n := range notifications {
+		h.incoming <- n
+	}
+}
+
+func (h *notificationHandler) Configure(apiConfig Config) {
+	h.configChan <- apiConfig
+}
+
+func (h *notificationHandler) Quit() {
+	h.quit <- true
+}
+
+func (h *notificationHandler) work() {
+	for {
+		select {
+		case n := <-h.incoming:
+			if h.queueSize >= maxQueueSize {
+				h.Logger.LogError(
+					fmt.Sprintf("Notification queue full (%v), dropped notification", h.queueSize))
+				continue
+			}
+			h.queueSize++
+			h.queue <- n
+
+		case n := <-h.queue:
+			h.queueSize--
+			h.processNotification(n)
+
+		case apiConfig := <-h.configChan:
+			h.Config = apiConfig
+
+		case <-h.quit:
+			return
+		}
+	}
+}
+
+func (h *notificationHandler) processNotification(n *remote.Notification) {
+	message := ""
+	changed := false
+	prior, err := h.EventStore.LoadUserEvent(n.Subscription.CreatorID, n.Event.ID)
+	switch err {
+	case kvstore.ErrNotFound:
+		changed = true
+		message = h.formatNewEventNotification(n)
+		prior = &store.Event{}
+	case nil:
+		changed, message = h.formatUpdatedEventNotification(n, prior)
+	default:
+		return
+	}
+	if !changed {
+		return
+	}
+
+	err = h.Poster.PostDirect(n.SubscriptionCreatorMattermostUserID, message, "")
+	if err != nil {
+		h.Logger.LogInfo("Failed to post notification message: " + err.Error())
+	}
+
+	prior.Remote = n.Event
+	err = h.EventStore.StoreUserEvent(n.Subscription.CreatorID, prior)
+	if err != nil {
+		return
+	}
+
+	h.Logger.LogDebug("Processed event notification",
+		"MattermostUserID", n.SubscriptionCreatorMattermostUserID,
+		"SubsriptionID", n.SubscriptionID,
+		"Message", message)
+	return
+}
+
+func (h *notificationHandler) formatUpdatedEventNotification(n *remote.Notification, prior *store.Event) (bool, string) {
+	priorFields := eventFields(prior.Remote)
+	newFields := eventFields(n.Event)
+	changed, added, updated, deleted := fields.Diff(priorFields, newFields)
+	if !changed {
+		return false, ""
+	}
+
+	message := fmt.Sprintf("Updated event: [%s](%s)\n", n.Event.Subject, n.Event.Weblink)
+	for _, k := range added {
+		message += fmt.Sprintf("- %s: %s\n", k, newFields[k].Strings())
+	}
+	for _, k := range updated {
+		message += fmt.Sprintf("- %s: ~~%s~~ \u2192 %s\n",
+			k, priorFields[k].Strings(), newFields[k].Strings())
+	}
+	for _, k := range deleted {
+		message += fmt.Sprintf("- %s: ~~%s~~\n", k, newFields[k].Strings())
+	}
+
+	h.Logger.LogDebug("Processed event notification",
+		"MattermostUserID", n.SubscriptionCreatorMattermostUserID,
+		"SubsriptionID", n.SubscriptionID,
+		"Message", message)
+	return true, message
+}
+
+func (h *notificationHandler) formatNewEventNotification(n *remote.Notification) string {
+	message := ""
+	if n.ChangeType == "crated" {
```

Comment:
I'm thinking this is a typo

```suggestion
	if n.ChangeType == "created" {
```

Reply:
LOL, I was wondering why it was not displaying the right headline!